husky_ws/src/husky/husky_navigation/config/nav2_params.yaml

```diff
@@ -125,6 +125,7 @@ controller_server:
     progress_checker_plugin: "progress_checker"
     goal_checker_plugins: ["general_goal_checker"] # "precise_goal_checker"
     controller_plugins: ["FollowPath"]
+    odom_topic: /odometry/filtered
```

Comment:
I'm uncertain if there's documents available for ROS2, but the document for ROS1 is quite comprehensive. You can refer to this [website](https://docs.ros.org/en/melodic/api/robot_localization/html/state_estimation_nodes.html) for more information.

The `ekf_node` is an implementation of the [Extended Kalman Filter](https://en.wikipedia.org/wiki/Extended_Kalman_filter), which is highly beneficial for sensor fusion tasks. However, comprehending its workings completely can be challenging. While I once had a comprehensive understanding of it, I find that I've forgotten some details over time.

Reply:
Thanks for the information. I seem to overlooked the topic name documented in its ROS 1 documentation: https://docs.ros.org/en/noetic/api/robot_localization/html/state_estimation_nodes.html#published-topics

The ROS 2 documentation for robot_localization doesn't seem to exist yet, since Nav2 also refers to the ROS 1 docs: https://navigation.ros.org/setup_guides/odom/setup_odom.html#robot-localization-demo

## Local/Global costmap misalignment

I'm asking this question because I suspect that the `ekf_node` might not be performing sensor fusion as intended. This hypothesis is due to the local costmap and global costmap misalignment mentioned previously. (Your recent fix mitigated the issue, but it still happen sometimes).

https://github.com/j3soon/ros2-agv-essentials/assets/20457146/6dd46462-1bbd-44cb-bd41-5a1d15f5d014

I've used the `rqt_graph` and `ros2 topic` tool and didn't see `ekf_node` using the IMU data.

![image](https://github.com/j3soon/ros2-agv-essentials/assets/20457146/8ed53fb5-c86d-43ea-85d2-59f055741fad)

![image](https://github.com/j3soon/ros2-agv-essentials/assets/20457146/c7189604-96c4-4856-aff2-4b49852f1d14)

This is due to incorrect gazebo IMU plugin topic name. Details of this fix is provided in the first commit below.

## Husky model and odom mismatch

After making `ekf_node` fuse IMU data, the odom pose will diverge from husky model pose.

![image](https://github.com/j3soon/ros2-agv-essentials/assets/20457146/525e44c0-68f7-4db2-845b-c1eae16bc6c2)

This can be easily fixed by modifying the RVIZ file in the second commit below.

## Large Y error due to IMU fusion

As simulation time passes, the Y error becomes larger and larger.

![image](https://github.com/j3soon/ros2-agv-essentials/assets/20457146/57cf3378-bed8-4243-bffa-f2548b2e4efc)

This can be fixed by not using Y data from the IMU as described in the third commit below.

https://github.com/j3soon/ros2-agv-essentials/assets/20457146/d2611c52-026a-4c32-8187-e9e21a068953

## Variance explosion

If the simulation runs for a longer time, the variance of odom estimation still explodes due to not having absolute data.

![image](https://github.com/j3soon/ros2-agv-essentials/assets/20457146/e905126f-1725-492e-ac59-08d0bb723458)

This can be fixed by fusing pose data provided by the slam_toolbox as mentioned in the fourth commit below. However, this causes jumps in odom estimation. I haven't spent time on investigating ways to mitigate this yet.

https://github.com/j3soon/ros2-agv-essentials/assets/20457146/f5df81a4-94e8-4728-9102-c6185ae2b92f

Reply:
Firstly, I want to express my gratitude for the fixes, especially for the commit addressing the topic issue. I believe it resolves an issue that has been bothering me for quite some time. Specifically, while addressing the previous issue, I attempted to mitigate the localization error by reducing the rotational speed. However, I noticed that the velocity smoother wasn't functioning correctly. At that time, I postponed addressing this issue to the future, but unfortunately, I forgot to resolve it in the end.

---

For directly fusing the pose from `slam_toolbox` into the EKF, I believe it's not a recommended practice due to the jumps in the data, as you mentioned. There is an potential solution:

Enable the config `dynamic_process_noise_covariance` to `true` in the localization.yaml file. This adjustment should help mitigate the issue by allowing the system to adapt its noise covariance dynamically. Remember to disable the fusion of pose in the ekf_node when you want to try this solution. [Reference](https://docs.ros.org/en/melodic/api/robot_localization/html/state_estimation_nodes.html#dynamic-process-noise-covariance)

---

Just a quick note. We should only integrate the global data into the ekf_node, as described below:

Add an additional EKF node to manage the fusion of pose data. Here's how it works: utilize a local EKF node to fuse odometry and IMU data, then employ a global EKF node to fuse the data from the local EKF node and the pose. This approach promises a more robust fusion strategy. You can find the concept explained [here](https://docs.ros.org/en/melodic/api/robot_localization/html/integrating_gps.html#using-navsat-transform-node). However, this strategy may not be applicable to solve the issue here, as we're relying on slam_toolbox to publish the tf map -> odom.

Reply:
Thanks for the information!

Seems like the correct way to go is removing the global pose info and use the `dynamic_process_noise_covariance`. Are you interested in giving it a try? If so, please feel free to make additional commits and remove the global pose information in my last commit.